### PR TITLE
blk/aio: fix long batch (64+K entries) submission.

### PR DESCRIPTION
**Pull request by ifed01**
This is a revival of https://github.com/ceph/ceph/pull/44065
Includes some fixes and UT.

Fixes: https://tracker.ceph.com/issues/64966

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ifed01** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test api

File: src/test/objectstore/store_test.cc [hunk 1]
```diff
@@ -7541,6 +7541,102 @@ TEST_P(StoreTestSpecificAUSize, BlobReuseOnOverwrite) {
   }
 }
```
Comment (aclamk) [inline src/test/objectstore/store_test.cc:7543]: Excellent test.

File: src/blk/aio/aio.cc [hunk 2]
```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
   int r;
 
   aio_iter cur = begin;
-  struct aio_t *piocb[aios_size];
```
Comment (aclamk) [inline src/blk/aio/aio.cc:28]: The problem, as I understand it, is very simple: `aios_size` got an overflow.
I think solution should be to either:
1) change `uint16_t` to `uint32_t` (preferred)
2) count elements between begin and end; use `alloca` to allocate piocb.

The only benefit of current solution is that it does minimize stack allocation, but at cost of added complexity. I do not think its worth it.
Note that for every entry in `piocb` we will be reading at least 4K from the drive. I think we can afford having entire `piocb` on stack.

Comment (ifed01) [reply src/blk/aio/aio.cc:28]: @aclamk - you might be missing one more aspect - amount of chunks (and hence amount IOs issued in a single batch) in an object is determined (to some degree)  by cluster's users. So it becomes possible for them to use "bad" objects which could DDOS OSDs or do something inappropriate. 
And this patch provides a sort of protection against that - at aio level at least..

File: src/blk/aio/aio.cc [hunk 3]
```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+    if ((cur->n_aiocb == 1) {
       // TODO: consider batching multiple reads together with lio_listio
-      piocb[done]->aio.aiocb.aio_sigevent.sigev_notify = SIGEV_KEVENT;
-      piocb[done]->aio.aiocb.aio_sigevent.sigev_notify_kqueue = ctx;
-      piocb[done]->aio.aiocb.aio_sigevent.sigev_value.sival_ptr = piocb[done];
-      r = aio_read(&piocb[done]->aio.aiocb);
+      cur->aio.aiocb.aio_sigevent.sigev_notify = SIGEV_KEVENT;
+      cur->aio.aiocb.aio_sigevent.sigev_notify_kqueue = ctx;
+      cur->aio.aiocb.aio_sigevent.sigev_value.sival_ptr = &(*cur);
+      r = aio_read(&cur->aio.aiocb);
```
Comment (aclamk) [inline src/blk/aio/aio.cc:56]: 'aio_read'? How on earth POSIXAIO could have ever worked with ceph? I refuse to believe that there were never single aio request to write one chunk.
Proposal:
Lets replace implementation here with `#error Implementation removed.`.

File: src/blk/aio/aio.cc [hunk 4]
```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+      cur->priv = priv;
+      piocb[itemCount] = &(*cur);
+      ++itemCount;
+      ++cur;
+    }
+    int toSubmit = itemCount - itemCount0;
+    r = io_submit(ctx, toSubmit, (struct iocb**)(piocb + itemCount0));
+    if (r >= 0 && r < toSubmit) {
+      r = -EAGAIN;
+      itemCount0 += (r > 0 ? r : 0);
```
Comment (aclamk) [inline src/blk/aio/aio.cc:46]: Easy task for an optimizer: `itemCount0 += 0;`, `as r==-EAGAIN` :)

File: src/blk/aio/aio.cc [hunk 5]
```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
-  }
-  ceph_assert(aios_size >= left);
+#if defined(HAVE_LIBAIO)
+  struct aio_t *piocb[max_iodepth];
+#endif
   int done = 0;
-  while (left > 0) {
+  int itemCount0 = 0; //used for LIBAIO only
+  int itemCount = 0;
+  while (cur != end || itemCount0 < itemCount) {
```
Comment (aclamk) [inline src/blk/aio/aio.cc:34]: s/ItemCount/ItemsPulled/
s/ItemCount0/ItemsPushed/
?
But this is obviously just my preference....

File: src/blk/aio/aio.cc (hunk 4 above)
Comment (ifed01) [reply src/blk/aio/aio.cc:46]: I believed I've fixed that... Let me check.

File: src/blk/aio/aio.cc (hunk 3 above)
Comment (ifed01) [reply src/blk/aio/aio.cc]: Yeah, that's funny. Replaced with aio_write()

File: src/blk/aio/aio.cc [hunk 6]
```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+    while (cur != end && pulled < max_iodepth) {
+      cur->priv = priv;
+      piocb[pulled] = &(*cur);
+      ++pulled;
+      ++cur;
+    }
+    int toSubmit = pulled - pushed;
+    r = io_submit(ctx, toSubmit, (struct iocb**)(piocb + pushed));
+    if (r >= 0 && r < toSubmit) {
+      pushed += (r > 0 ? r : 0);
```
Comment (rzarzynski) [inline src/blk/aio/aio.cc:45]: I don't see a point in the `(r > 0)` comparisons. As we are in the if's body, `r` must be already `0` or positive. However, the ternary operator for `r == 0` yields... `0` IIUC.

Comment (ifed01) [reply src/blk/aio/aio.cc]: right, my bad. thanks!

**Comment by aclamk** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test make check

**Comment by yuriw** [comment]
This PR is under test in [https://tracker.ceph.com/issues/65796](https://tracker.ceph.com/issues/65796).

**Comment by kamoltat** [comment]
RADOS APPROVED https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues65796

**Pull request merged**